### PR TITLE
Fix add ranges in MultiFileNameParser

### LINKED ISSUE #22597: DirectILLCollectData cannot handle complex run number ranges
**Original reporter:** Jacques Ollivier ILL.

### Expected behavior

The following script should execute: 
```python
# Initialise directories
mantid.config.appendDataSearchDir(
    '/net4/serdon/illdata/processed/182/in5/exp_TEST-2911/rawdata/')

# Where there is a vanadium
mantid.config.appendDataSearchDir(
    '/usr/illdata/181/in5/internalUse/rawdata/')

mantid.config.appendDataSearchDir(
    '/net4/serdon/illdata/processed/182/in5/exp_TEST-2911/processed')

# Save output workspaces
ConfigService.Instance().setString('defaultsave.directory',
                                   '/net4/serdon/illdata/processed/182/in5/exp_TEST-2911/processed/data/mantid_nxs/')


sample = '168383-168384+168389-168390+168416-168420'

DirectILLCollectData(Run=sample,
                        OutputWorkspace='output-raw',
                        OutputEPPWorkspace='Sample-epp')
```

### Actual behavior

Mantid throws:

```
ValueError: Invalid value for property Run (list of str lists) "168383-168384+168389-168390+168416-168420": Unable to find file matching the string "168383-168384+168389-168390+168416-168420", even after appending suggested file extensions.
  at line 25 in 'New script'
  caused by line 1080 in '/users/soininen/mantid-build-Default/bin/mantid/simpleapi.py'
  caused by line 1029 in '/users/soininen/mantid-build-Default/bin/mantid/simpleapi.py'
  caused by line 1013 in '/users/soininen/mantid-build-Default/bin/mantid/simpleapi.py'
```

### Steps to reproduce the behavior

Try to `Load` any data using similar ranges as above.

### Platforms affected

All

### PR DESCRIPTION
**Pull request by soininen**
This PR enables handling of complex add ranges in `MultiFileNameParser` and consequently in `Load`, `LoadAndMerge` and similar algorithms that accept ranges of run numbers.

Previously, the parsing of an input string was rigidly based on regular expression matching which did not take into account more advanced expressions like `1+2-6+8-10+15`. Now, the tokens in the input string are further broken into smaller pieces. Only the individual pieces are matched against regexps which makes is possible to parse more complex patterns.

**Report to:** Jacques Ollivier ILL.

**To test:**

Try to `Load` or `LoadAndMerge` some files using different combinations of run number patterns such as `1+4-5+10-15:2` and check that the correct runs get loaded and merged.

Fixes #22597.

Does this update require release notes?
- [x] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/src/MultiFileNameParser.cpp [hunk 1]
```diff
@@ -620,11 +596,11 @@ std::vector<std::vector<unsigned int>> generateRange(unsigned int from,
     while (currentRun >= to) {
       if (addRuns) {
         if (runs.empty())
-          runs.push_back(std::vector<unsigned int>(1, currentRun));
+          runs.emplace_back(std::vector<unsigned int>(1, currentRun));
```
Comment (josephframsay) [inline Framework/Kernel/src/MultiFileNameParser.cpp]: I think you can replace this and 603 with `runs.emplace_back(1, currentRun)` which potentially saves you a copy

File: Framework/Kernel/src/MultiFileNameParser.cpp [hunk 2]
```diff
@@ -461,96 +455,79 @@ namespace // anonymous
...
  * @param token      :: the token to parse.
- *
- * @returns the newly parsed runs appended to the previously parsed runs.
- * @throws std::runtime_error if
  */
-std::vector<std::vector<unsigned int>> &
-parseToken(std::vector<std::vector<unsigned int>> &parsedRuns,
-           const std::string &token) {
-  // Tokenise further, on plus, minus or colon.
+void parseToken(std::vector<std::vector<unsigned int>> &parsedRuns,
```
Comment (josephframsay) [inline Framework/Kernel/src/MultiFileNameParser.cpp]: Line 451 isn't correct any more - you can't use this method with `std::accumulate`. What was the reason behind making the function non `accumulatable`-able?

**Review by josephframsay — CHANGES_REQUESTED** [review]
Changes are mostly nice, just a couple of minor comments. Is there any documentation for the meaning and order of precedence of the various operators in the run string?

**Comment by soininen** [comment]
@josephframsay The syntax is documented here: http://docs.mantidproject.org/nightly/api/python/mantid/api/MultipleFileProperty.html#mantid.api.MultipleFileProperty

File: Framework/Kernel/src/MultiFileNameParser.cpp (hunk 2 above)
Comment (soininen) [reply Framework/Kernel/src/MultiFileNameParser.cpp]: To me the way `std::accumulate` was (ab)used with this function was confusing. What is needed is just a simple loop over the tokens, see lines 132-134. I just forgot to update the doxygen so its good that you noticed.

**Comment by josephframsay** [comment]
@soininen would it be possible to add something about the precedence of each operator, as this isn't really made clear in the docs?

**Review by josephframsay — APPROVED** [review]

**Comment by martyngigg** [comment]
@soininen It was noted that this should be reported back to Jacques Ollivier at ILL.

**Pull request merged**